File: greedy/1353_maximum_number_of_events_that_can_be_attended/1353_maximum_number_of_events_that_can_be_attended.py
import heapq

class Solution(object):
    def maxEvents(self, events):
        ans = 0
        end = list()
        events = sorted(events, reverse=True)
        print events
        for i in range(1, 100010, 1):
            while events and events[-1][0] == i:
                heapq.heappush(end, events.pop()[1])
            while end and end[0] < i:
                heapq.heappop(end)
            if end:
                heapq.heappop(end)
                ans += 1
        return ans

events = [[1,2],[2,3],[3,4]]
res = Solution().maxEvents(events)
print(res)

File: greedy/1353_maximum_number_of_events_that_can_be_attended/1353_maximum_number_of_events_that_can_be_attended.go
package main

import (
	"fmt"
	"sort"
)

func maxEvents(events [][]int) int {
	var (
		i, j, k, maxEnd int
		length          = len(events)
		visitedDays     []bool
	)
	if length == 0 {
		return 0
	}

	maxEnd = events[0][1]
	sort.Slice(events, func(i, j int) bool {
		if events[i][1] > maxEnd {
			maxEnd = events[i][1]
		}
		if events[i][1] == events[j][1] {
			return events[i][0] > events[j][0]
		}
		return events[i][1] < events[j][1]
	})
	visitedDays = make([]bool, maxEnd+1)

	for i = 0; i < length; i++ {
		for j = events[i][0]; j <= events[i][1]; j++ {
			if !visitedDays[j] {
				visitedDays[j] = true
				break
			}
		}
	}

	for i = 0; i <= maxEnd; i++ {
		if visitedDays[i] {
			k++
		}
	}
	return k
}

func main() {
	events := [][]int{{1, 2}, {2, 3}, {3, 4}}
	res := maxEvents(events)
	fmt.Println(res)
}
